pkg/nodeagent/pod_controller.go

```diff
@@ -237,101 +244,152 @@ func buildRmdWorkload(pod *corev1.Pod) ([]*intelv1alpha1.RmdWorkload, error) {
 		}
 
 		rmdWorkload := &intelv1alpha1.RmdWorkload{}
-		// Create workload name. Naming convention: "<pod-name>-rmd-workload-<container-name>"
+		//Create workload name. Convention: "<pod-name>rmd-workload-<container-name>"
 		podName := string(pod.GetObjectMeta().GetName())
 		rmdWorkloadName := fmt.Sprintf("%s%s%s", podName, rmdWorkloadNameConst, container.Name)
 		podNamespace := pod.GetObjectMeta().GetNamespace()
 		if podNamespace == "" {
 			podNamespace = defaultNamespace
 		}
-
 		rmdWorkloadNamespacedName := types.NamespacedName{
 			Name:      rmdWorkloadName,
 			Namespace: podNamespace,
 		}
 
 		rmdWorkload.SetName(rmdWorkloadNamespacedName.Name)
 		rmdWorkload.SetNamespace(rmdWorkloadNamespacedName.Namespace)
-
-		maxCache, err := getMaxCache(&container)
-		if err != nil {
-			return nil, err
-		}
-		rmdWorkload.Spec.Rdt.Cache.Max = maxCache
-		rmdWorkload.Spec.CoreIds = coreIDs
+		rmdWorkload.Spec.Rdt.Cache.Max = containerInfo.maxCache
+		rmdWorkload.Spec.CoreIds = containerInfo.coreIDs
 		rmdWorkload.Spec.Nodes = make([]string, 0)
 		rmdWorkload.Spec.Nodes = append(rmdWorkload.Spec.Nodes, pod.Spec.NodeName)
 
 		workloadData := pod.GetObjectMeta().GetAnnotations()
-		for field, data := range workloadData {
-			if !strings.HasPrefix(field, container.Name) {
-				continue
-			}
+		policy, minCache, mbaPercentage, mbaMbps, pstateRatio, pstateMonitoring := getAnnotationInfo(workloadData, container)
 
-			switch {
-			case strings.HasSuffix(field, policyConst):
-				if data != "" {
-					rmdWorkload.Spec.Policy = data
-				}
-			case strings.HasSuffix(field, cacheMinConst):
-				minCache, err := strconv.Atoi(data)
-				if err != nil {
-					continue
-				}
-				rmdWorkload.Spec.Rdt.Cache.Min = minCache
-			case strings.HasSuffix(field, mbaPercentageConst):
-				mbaPercentage, err := strconv.Atoi(data)
-				if err != nil {
-					continue
-				}
-				rmdWorkload.Spec.Rdt.Mba.Percentage = mbaPercentage
-			case strings.HasSuffix(field, mbaMbpsConst):
-				mbaMbps, err := strconv.Atoi(data)
-				if err != nil {
-					continue
-				}
-				rmdWorkload.Spec.Rdt.Mba.Mbps = mbaMbps
-			case strings.HasSuffix(field, pstateRatioConst):
-				if data != "" {
-					rmdWorkload.Spec.Plugins.Pstate.Ratio = data
-				}
-			case strings.HasSuffix(field, pstateMonitoringConst):
-				if data != "" {
-					rmdWorkload.Spec.Plugins.Pstate.Monitoring = data
-				}
-			}
+		rmdWorkload.Spec.Policy = policy
+		rmdWorkload.Spec.Rdt.Cache.Min = minCache
+		rmdWorkload.Spec.Rdt.Mba.Percentage = mbaPercentage
+		rmdWorkload.Spec.Rdt.Mba.Mbps = mbaMbps
+		rmdWorkload.Spec.Plugins.Pstate.Ratio = pstateRatio
+		rmdWorkload.Spec.Plugins.Pstate.Monitoring = pstateMonitoring
 
-		}
 		rmdWorkloads = append(rmdWorkloads, rmdWorkload)
 	}
 	return rmdWorkloads, nil
 }
 
+func getAnnotationInfo(workloadData map[string]string, container corev1.Container) (policy string, minCache int, mbaPercentage int, mbaMbps int, pstateRatio string, pstateMonitoring string) {
```

Comment:
You are on the right track by extracting this logic out into its own function. However, I think we can improve it a lot more by passing a pointer to the RmdWorkload object to the function and letting the function be responsible for mutation of the object instead of returning all that data to the larger `buildRmdWorkload` function.

In doing so it should also be renamed to something more suitable. Also, this function can also take the pod object and and the container name ans parameters (the full container is unnecessary). This way we can add GetAnnotations() logic inside this function too.

EG:
````
func getAnnotationInfo(rmdWorkload *intelv1alpha1.RmdWorkload, pod *corev1.Pod, containerName string){
  . . . 
  workloadData := pod.GetObjectMeta().GetAnnotations()
  . . . 
}
````